mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -48,15 +48,416 @@ using namespace mlir::miopen;
 // 2G ,INT MAX Value = 2147483647, use 2147483648 as offset and buffer
 // store do nothing
 const int twoGB = 2147483647;
+//===----------------------------------------------------------------------===//
+// Utility function to compute index diff map.
+//===----------------------------------------------------------------------===//
+inline void computeIndexDiffMap(
+    OpBuilder &b, Location loc, const SmallVector<Value, 8> &upperIndicesDiff,
+    const DictionaryAttr &transformMetadata, const AffineMap &transform,
+    const SmallVector<Value, 8> &lowerIndicesOriginal,
+    SmallVector<Value, 8> &lowerIndicesDiff,
+    SmallVector<Value, 8> &lowerIndicesUpdated) {
+  auto zeroConstantI32Op =
+      b.create<ConstantIntOp>(loc, 0, b.getIntegerType(32));
+  auto oneConstantI32Op = b.create<ConstantIntOp>(loc, 1, b.getIntegerType(32));
+
+  // Obtain the shape of lower level memref.
+  ArrayAttr lowerLayerShape =
+      transformMetadata.get("lower_layer_bounds").template cast<ArrayAttr>();
+
+  // Input:
+  // - upper_diff
+  // - upper_indices_original
+  // - upper_layer_bounds
+  // - lower_indices_original
+  // - lower_layer_bounds
+  // - F : a vector of functions mapping upper level dimensions to lower level
+  // dimensions.
+  // - G : metadata of F
+  //
+  // Output:
+  // - lower_diff
+  // - lower_indices_updated
+  //
+  // For each transform g specified in G:
+  //   Let P be the upper dimensions used by g.
+  //   Let Q be the lower dimensions used by g.
+  //
+  //   Switch g:
+  //     Case Pad :
+  //       |P| = |Q|
+  //       For each i in P, and its counterpart j in Q
+  //         lower_diff[j] = upper_diff[i]
+  //
+  //     Case PassThrough :
+  //       |P| = |Q|
+  //       For each i in P, and its counterpart j in Q
+  //         lower_diff[j] = upper_diff[i]
+  //
```

Comment:
`Slice` used by `conv2d_bwd_data` needs to be considered here.

Reply:
Applied a dirty hack for `conv2d_bwd_data`. `Slice` really builds an identity map which would be optimized away by MLIR by default. Temporarily added a dirty fake map `(d0) -> (d0 + d0 + 1) / 2` to prevent from being optimized away. Need to find a better approach.

+ @asroy / @ltqin / @asleepzzz for awareness.

Reply:
Remove all the hacks.